### PR TITLE
[now-cli] Use XDG standard instead of writing to home directory

### PR DESCRIPTION
**Pull request by mstruebing**
references: #1909 

At first: we should decide if this is `config` or `data` regarding the XDG-specification.
My guess would be data because it holds access tokens and nothing really config related.
But I would be willing to use config if the majority says so.

cc @leo @rivy 

@rivy the creation of the directory is handled elsewhere, in `packages/now-cli/src/index.js`, so I don't need to take care of that, also this shouldn't be part of this file to not mix responsibilities imo.

### CODE REVIEW

**Comment by leo** [comment]
Thanks for moving over the PR!

**Comment by leo** [comment]
Like I said on the previous PR, I think this is a great idea.

Let's check and merge it, @AndyBitz @TooTallNate.

**Review by TooTallNate — CHANGES_REQUESTED** [review]
Code looks pretty good except for a few typos in comments.

But also, this will need at least a test case that ensures the proper directory is being used for the configuration.

**Review by rivy — CHANGES_REQUESTED** [review]

File: packages/now-cli/src/util/config/global-path.ts [hunk 1]
```diff
@@ -16,11 +40,14 @@ const getNowDir = () => {
 
   const customPath = args['global-config'];
 
-  if (!customPath) {
-    return path.join(homedir(), '.now');
-  }
-
-  return path.resolve(customPath);
+  // We use the customPath if it is available,
+  // otherwise the config at the users homedirectory if it is present
```
[the remark this answers is not in the document]
Comment (mstruebing) [reply packages/now-cli/src/util/config/global-path.ts]: Fixed :)

File: packages/now-cli/src/util/config/global-path.ts [hunk 2]
```diff
@@ -1,12 +1,36 @@
...
+const isDirectory = (path: string): boolean => {
+  try {
+    return fs.lstatSync(path).isDirectory();
+  } catch (_) {
+    // We don't care which kind of error occured, it isn't a directory anyway.
+    return false;
+  }
+};
+
+// Returns in which directory the cofig should be present
```
[the remark this answers is not in the document]
Comment (mstruebing) [reply packages/now-cli/src/util/config/global-path.ts]: Fixed :)

File: packages/now-cli/src/util/config/global-path.ts [hunk 3]
```diff
@@ -1,12 +1,36 @@
...
+const homeConfigPath: string = path.join(homedir(), '.now');
+
+// The `xdgConfigPath` is the configuration path which is based on the XDG standard.
+// the old legacy path is used as a fallback
+const xdgConfigPath: string = path.join(
+  head(xdg.dataDirs()) || homeConfigPath,
+  'now'
+);
+
+// Returns wether a directory exists
```
[the remark this answers is not in the document]
Comment (mstruebing) [reply packages/now-cli/src/util/config/global-path.ts]: Fixed :)

File: packages/now-cli/package.json [hunk 4]
```diff
@@ -198,5 +198,10 @@
     "which-promise": "1.0.0",
     "write-json-file": "2.2.0",
     "yarn": "1.17.3"
+  },
+  "dependencies": {
```
[the remark this answers is not in the document]
Comment (mstruebing) [reply packages/now-cli/package.json]: Fixed :)

File: packages/now-cli/src/util/config/global-path.ts [hunk 5]
```diff
@@ -1,12 +1,36 @@
...
+import head from 'lodash.head';
 
-const getNowDir = () => {
+// The `homeConfigPath` is the legacy configuration path located in the users home directory.
+const homeConfigPath: string = path.join(homedir(), '.now');
+
+// The `xdgConfigPath` is the configuration path which is based on the XDG standard.
+// the old legacy path is used as a fallback
+const xdgConfigPath: string = path.join(
+  head(xdg.dataDirs()) || homeConfigPath,
```
[the remark this answers is not in the document]
Comment (mstruebing) [reply packages/now-cli/src/util/config/global-path.ts]: I've removed the lodash dependency and used your suggested change.

File: packages/now-cli/src/util/config/global-path.ts [hunk 6]
```diff
@@ -16,11 +40,14 @@ const getNowDir = () => {
...
-
-  return path.resolve(customPath);
+  // We use the customPath if it is available,
+  // otherwise the config at the users homedirectory if it is present
+  // and if nothing of this is there we use the XDG-standard
+  return (
+    (customPath && path.resolve(customPath)) ||
+    (isDirectory(homeConfigPath) && homeConfigPath) ||
+    xdgConfigPath
+  );
```
[the remark this answers is not in the document]
Comment (mstruebing) [reply packages/now-cli/src/util/config/global-path.ts:45]: I've tried to satisfy your concerns, is it how you like now?

**Comment by mstruebing** [comment]
Is the failing CI part of my PR? I hope I didn't touch anything which has to do with wildcard aliases.

File: packages/now-cli/src/util/config/global-path.ts [hunk 7]
```diff
@@ -15,12 +31,18 @@ const getNowDir = () => {
...
-  }
-
-  return path.resolve(customPath);
+  // customPath is the preferred location
+  // the legacy home directory config path is the second most wanted location
+  // otherwise the first matching xdg-config-path is used which already exists
+  // at last the first best xdg-config-path is used
+  return (
+    (customPath && path.resolve(customPath)) ||
+    possibleConfigPaths.filter(configPath => isDirectory(configPath))[0] ||
```
Comment (AndyBitz) [inline packages/now-cli/src/util/config/global-path.ts]:
```suggestion
    possibleConfigPaths.find(configPath => isDirectory(configPath)) ||
```

I guess we can just use the first one that is a directory

Comment (mstruebing) [reply packages/now-cli/src/util/config/global-path.ts]: Good point, I somehow always don't remember the `find`-function. Thank you.

**Review by TooTallNate — APPROVED** [review]

**Comment by TooTallNate** [comment]
@mstruebing Thank you for the PR! Merged.

**Comment by mstruebing** [comment]
WHOOP WHOOP, thanks for merging :) :+1:

**Pull request merged**